flow_apps/Flow-Based Knowledge/force-app/main/default/classes/KnowledgeRecommendationsGenerator.cls

```diff
@@ -0,0 +1,91 @@
+{
+    @InvocableMethod
+    public static List<List<Recommendation>> generateKnowledgeRecommendations(List<Requests> requests)
+    {
+        List<List<Recommendation>> recommendationsResult = new List<List<Recommendation>>();
+        List<Recommendation> recommendations = new List<Recommendation>();
+
+        String searchString = requests[0].searchString;
+        String languageLocale = requests[0].languageLocale;
+        String articleType = requests[0].articleType;
+        String topic = requests[0].topic;
+        String publicationStatus = requests[0].publicationStatus;
+        Integer resultLimit = requests[0].resultLimit;
+        
+        if(resultLimit == null)
+        {
+            resultLimit = 5;
+        }
+
+        String queryString = 'FIND {' + searchString + '} IN ALL FIELDS RETURNING KnowledgeArticleVersion(';
+        queryString += 'id, title WHERE Language = :languageLocale';
+        /* Commented, because such field doesn't exists on KnowledgeArticleVersion object
+        if(articleType != null)
+        {
+            queryString += ' AND ArticleType = :articleType';
+        }
+        */
+        if(topic != null)
+        {
+            queryString += ' AND Title = :topic';
+        }
+
+        if(publicationStatus != null)
+        {
+            queryString += ' AND PublishStatus = :publicationStatus';
+        }
+
+        queryString += ') WITH SNIPPET (target_length=120) LIMIT ' + resultLimit;
+        Search.SearchResults searchResults = Search.find(queryString);
+        List<Search.SearchResult> articleList = searchResults.get('KnowledgeArticleVersion');
+        
+        for(Search.SearchResult searchResult : articleList)
+        {
+            Recommendation recommendation = new Recommendation();
+	        KnowledgeArticleVersion article = (KnowledgeArticleVersion) searchResult.getSObject();
+            
+            recommendation.ExternalId = article.Id;
+            recommendation.Name = article.Title;
+            recommendation.Description = searchResult.getSnippet();
+            recommendation.ActionReference = 'Display_Knowledge_Article';
+
+            recommendations.add(recommendation);
+        }
+
+        recommendationsResult.add(recommendations);
+        return recommendationsResult;
+    }
+
+    @AuraEnabled
+    public static List<Recommendation> executeNBAFlow(String strategyName, String contextRecordId, String searchString) {
+        Flow.Interview curInterview = Flow.Interview.createInterview(strategyName, new Map<String,Object>{'recordId' =>contextRecordId, 'searchString' => searchString });
+        curInterview.start();
+        System.debug('searchString' + searchString);
+        System.debug(curInterview.getVariableValue('outputRecommendations'));
+        return (List<Recommendation>)curInterview.getVariableValue('outputRecommendations');
+    }
+
+    public class InvocableErrorException extends Exception {}
+
+    public class Requests 
```

Comment:
Just as you have a class here for Requests, you should use an inner class for Responses, even though this class currently only returns a single List of Recommendations. Doing this now makes it easy to later add additional outputs without affecting the signature of the class.

Reply:
resolved it